### ISSUE TITLE
Cluster Discovery for single Cache Cluster ID

### ISSUE THREAD

**Issue by AndresPineros** [maxbrunet/prometheus-elasticache-sd#23]
Hello,
First of all, thanks for this project. It is a great idea to be able to automatically fetch endpoints from cluster nodes automatically.

I'm having an issue. Currently I have several clusters with password, which means I'm forced to deploy one redis_exporter per each cluster, because redis_exporter only works with a single password for all the provided targets. This is fine, a bit more resource consumption but not a big deal. Because of that, I was trying to setup the prometheus-elasticache-sd to work with a single cache cluster (--elasticache.cache-cluster-id) but I just realized that the AWS API is returning each redis node as its own cluster. The reason is that we actually have ReplicationGroups for redis, which creates a cache-cluster for each node (apparently).

Let's assume I have an elasticache with the following properties:
- name: my-elasticache-redis
- nodes: 2 (1 primary, 1 read replica)
- shards: 1

I'd expect the prometheus-elasticache-sd to receive this name "my-elasticache-redis" and return the endpoints from the 2 nodes. But it will say that there is no cache cluster named "my-elasticache-redis". When I query the cache clusters with the AWS API, I see 2 cache-clusters: "my-elasticache-redis-001" and "my-elasticache-redis-002"... Which defeats the purpose of autodiscovering the nodes. If I set `--elasticache.cache-cluster-id=my-elasticache-redis-001` it works but only registers the endpoint from one of the nodes, not both.

What are your thoughts around adding support for ReplicationGroups?

**Comment by maxbrunet** [comment]
Hi @AndresPineros, thank you for opening this issue, yes, this API is little limited. Replication groups are only an option for Redis, and I would like this service discovery to remain as generic as possible.

For now, I had in mind that if any filtering is needed beyond what the [ElasticCache API](https://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_DescribeCacheClusters.html) offers, relabeling should be used, for example:

```yaml
- source_labels: [__meta_elasticache_replication_group_id]
  regex: my-elasticache-redis
  action: keep
```

Would the above work for you?

I know this is not the most optimal, but it allows to keep things simple. If AWS ever adds a `filters` parameter like the [EC2 API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html), I would be happy to expose it via a CLI flag.

**Comment by AndresPineros** [comment]
Hey @maxbrunet, yes, this solution would definetely work for me. 

Unfourtunately, I'm working in a Kubernetes environment and realized that the file service discovery would be extremely cumbersome when working with multiple redis environments and the Prometheus Operator. I know this isn't related to this issue but I'll explain anyways. I'd have to setup an emptyDir between prometheus and each redis-discovery instance to share the file_sd, which isn't maintainable. If this project aims to integrate with Kubernetes in the future, the discovered Redis endpoints could be written into a Prometheus Probe CRD instead of a file. Prometheus can discover the Probe CRD and send requests to the specified targets. This would be the most maintainable solution IMO.

Anyways, awesome project, thanks for making it public :)

**Comment by maxbrunet** [comment]
Hey @AndresPineros, I use this SD with the Prometheus Operator. For the `emptyDir`, you can re-use the `config-out` volume used by the operator, the output path could look like `/etc/prometheus/config_out/elasticache_<replication-groupd-id>_file_sd.yml`. And for the `file_sd` scrape config, the `spec.additionalScrapeConfigs` secret selector of the Prometheus CRD can be used.

If you use [prometheus-operator/kube-prometheus](https://github.com/prometheus-operator/kube-prometheus) (which I recommend), you can create something like a Jsonnet function which takes a list of replication group IDs as argument and loops over them to generate the needed config (depending on how many groups you have it could be a lot of sidecars, I am not sure if there are any limitations here).

I only have a single ElastiCache SD sidecar, and once everything is implemented in my Jsonnet library, I only need to pass a small config for each environment:

```jsonnet
{
  values+:: {
    prometheus+: {
      elastiCacheSD+: {
        enabled: true,
        matchTags: {
          // AWS tags used for filtering in `action: keep` relabeling rules
        }
      },
    }
  }
}
```

I might share that in this repository as a Jsonnet add-on when I have time.

**Comment by AndresPineros** [comment]
Yes, this would work but we're trying to avoid having "shared" resources, or traversal resources. Imagine this scenario: We need to deploy a new environment for another team in our company. They deploy everything, even the Redis ServiceDiscovery with the RedisExporter and the rest of their applications.

THEN they come back us and say "Hey, I don't see the Redis metrics in my Grafana". We realize we have to go to Prometheus and update the JSonnet function to add the new ReplicationGroup name. This means there are 2 steps to deploying.

In the scenario I described above with the Probe CRD, there's only 1 step. No need to modify the Prometheus CRD configuration. Each environment is decoupled from Prometheus using the Probe CRD. You'd just deploy the environment with the Redis ServiceDiscovery, and the ServiceDiscovery could create the Probe automatically, so now Prometheus would know how to hit the endpoints with no configs modified. The Probe is basically a file_sd that automatically configures Prometheus to hit the endpoints. I think ideally we should always try to have environments that can be independent from centralized resources, in this case the environment shouldn't depend on updating a Prometheus config.

We want to avoid the "step 1: do this. step 2: do that. step 3: ..." to make the whole stack work.

I do understand that writing the logic in this repo to create a Probe CRD would couple this really hard to K8s. That's why I proposed the adapter that is capable of transforming a file_sd into a Probe CRD on real time. This way this project remains decoupled from K8s and there's no need for multiple steps when deploying an environment with Redis Service Discovery.
